Enable worker to reconnect to gearman indefinitely

toli-belo opened the pull request:
In most cases, such as gearman's restart worker will attempt to reconnect to gearman indefinitely, however, in case of permanent net.OpError there was previously logic to self-terminate the worker, which I'm keeping here until we have a chance to review whether we still want it.

Commit:
Add agent reference to ErrorHandler args to allow specific connection recovery

Commit:
Chang disconnect logic to attempt reconnecting to server indefinitely

Commit:
Change Agent struct back to un-exported "agent"

Commit:
Change Agent member back to un-exported "agent"

Commit:
Removed outdated comment

Commit:
Rename agent.Addr to agent.addr since don't need it to be exported

Commit:
Add/update persistent reconnect behavior to client, worker

* Add persistent initial connection to worker, consolidate Connect() and Reconnect()
* Add persistent initial connection and re-connect behavior to client
* Replace NewConnected client constructor with a universal version which takes
   connection open/close handlers as arguments.
* Add io error handling in a few io places where it was missing

Commit:
Update readme with updated function name

ssmccoy commented on README.md:
````diff
@@ -106,7 +106,7 @@ go w.Work()
 
 ```go
-c, err := client.New("tcp4", "127.0.0.1:4730")
+c, err := client.NewNetClient("tcp4", "127.0.0.1:4730")
````
I'm not sure why you change the constructor name here, it just breaks client compatibility.

Is there a reason to?

toli-belo replied:
The idea behind the change was that in our case "NewNetClient" represents a rather specialized case, and it would be less confusing to public than using "New" should be reserved for a general/universal case. This became more apparent after replacing NetConnected with NewClient which has become the generalized version where connectivity is handled by lambdas.

ssmccoy replied:
I'm not sure that's a good reason to break the API

toli-belo replied:
renamed back to `New`

ssmccoy commented on client/client.go:
```diff
@@ -49,25 +53,46 @@ type Client struct {
+			}
+			conn, err = net.Dial(network, addr)
+			if err != nil {
+				time.Sleep(500 * time.Millisecond)
```
Why not just use `net.DialTimeout` ?

ssmccoy replied:
Please a) Use an exponential backoff, b) Do not hard code the timeout values (a constructor like `New` can specify reasonable defaults, but as with the `New` `NewConnected` split that existed earlier you should allow overriding this value and probably also its unit without changing the client code.

toli-belo replied:
Since we're expecting the gearman to come back online within a short amount of time I didn't see that we want the exponential backoff because imagine that recovery took a few minutes and our exponential backoff would have already increased the retry delay to perhaps range of minutes, which would further delay coming back online, so the longer it takes to bring gearman online the worse our complete recovery is affected exponentially. I would say we use exponential backoff for resources that have unknown recovery goal (those that might never come back online?)

toli-belo replied:
Also `net.DialTimeout` does allow to specify the delay between tries, it allows to set your own timeout for various steps needed to establish the connection (e.g DNS lookup), Tested this without time.sleep() and this made hundreds of attempts to reconnect within seconds. 

> DialTimeout acts like Dial but takes a timeout.
> 
> The timeout includes name resolution, if required. When using TCP, and the host in the address parameter resolves to multiple IP addresses, the timeout is spread over each consecutive dial, such that each is given an appropriate fraction of the time to connect.

ssmccoy commented on client/client.go:
```diff
@@ -49,25 +53,46 @@ type Client struct {
-// testing, though other use-cases can be imagined.
-func NewConnected(conn net.Conn) (client *Client) {
+/// handler_conn_close: optional
+func NewClient(handler_conn_close ConnCloseHandler,
```
Go convention is to use `camelCase` for private members, not underscores.

toli-belo replied:
fixed this

toli-belo replied:
restored `NewConnected` function

ssmccoy commented on client/client.go:
```diff
@@ -49,25 +53,46 @@ type Client struct {
+	handler_conn_open ConnOpenHandler) (client *Client) {
+
+	conn, err := handler_conn_open()
+	if err != nil {
```
Why not implement the timeout/retry and backoff _here_ instead of in the other location?

toli-belo replied:
It would have been implemented here if our Client wasn't compatible with non-TCP type of connection. Since seeing the need for net.Pipe I moved reconnect logic into the handler function.

ssmccoy commented on client/client.go:
```diff
@@ -87,6 +114,12 @@ func NewConnected(conn net.Conn) (client *Client) {
 }
 
+func (client *Client) getConn() *connection {
+	client.Lock()
```
Please do not add this lock, a lot of effort was put into removing the weird global locks in this code and replacing it with compare and swap. Your changes need to be resilient to that design decision because otherwise the lock (unless you hang onto it for the _duration of interacting with the connection_) is not actually doing you any good here.

jacksonrnewhouse replied:
+1 to scott, this doesn't seem to do anything, since returning client.conn is just passing a pointer back.

toli-belo replied:
The purpose of this lock is actually to prevent the readLoop from using a connection while the writeLoop has seen an error on that connection and already started reconnect.

ssmccoy commented on client/client.go:
```diff
@@ -137,47 +193,33 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 	// If it is unexpected error and the connection wasn't
 	// closed by Gearmand, the client should close the conection
 	// and reconnect to job server.
-	client.Close()
```
If you are removing this line, please remove the comment above.

toli-belo replied:
fixed

ssmccoy commented on client/client.go:
```diff
@@ -137,47 +193,33 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
+	log.Warningf("Closing connection to %v due to error %v, will reconnect...", client.addr, err)
 
-	conn, err := net.Dial(client.net, client.addr)
+	client.Lock()
```
Please do not add this lock, aforementioned applies here

toli-belo replied:
see my comment about the purpose of the Lock above.

ssmccoy commented on client/client.go:
```diff
@@ -137,47 +193,33 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 		return err
 	}
 
-	swapped := atomic.CompareAndSwapPointer(
```
Why would you replace CAS with a Mutex?

toli-belo replied:
CompareAndSwapPointer was removed only because I added use of lock(). Lock was added not because it's a substitute to CAS, instead the Lock() solution was chosen as a way to prevent the readLoop() and writeLoop() from using an old connection handle while the connection is being recovered in the reconnect() code initiated by either readLoop() or writeLoop(). As of the latest commit b19aec8, the other thread will wait for reconnection to complete and then proceed from there. This works in my testing so far, except for one kink with disconnect occurring right between sending command on a channel "outbound" and receiving a response on channel "expected"

toli-belo replied:
restored CAS with updated logic.

Commit:
Add a sense of ownership of which thread (read or write) is reconnecting

Problem: suppose readLoop() encountered an error on the connection,
it starts the reconnect process, but suppose at nearly the same time the
writeLoop() thread was using the same connection handle and also encountered
an error (since connection is broken). This fix is to prevent both of them attempting
to reconnect and have one of them claim the ownership of reconnect process
and have the other thread merely wait for it to complete and exit reconnect,
and then go on with using the new conn handle via getConn().

Commit:
Fix handle naming to camel case

Commit:
Remove orphaned comment

Commit:
Implement erroring out of orphaned waiters on chan expected by using channel close

Also changed couple of error events in agent and worker to cause reconnection.

Commit:
Some logging for investigation

Commit:
Make sure worker purges orphaned tasks when associated gearman dies

Commit:
Allow Write to bypass the agent lock

As result the previous Write() function becams identical to unexported write() so we export it.

Commit:
Rearrange agent locking and Grab() placement.

Make sure Grab is called after each successful execution.
NOTE: that it's also called when reconnections are made.

Commit:
Remove appscode/go/log, and remove got handle and got result messages

toli-belo commented on client/client.go:
```diff
@@ -8,14 +8,14 @@ import (
 	"time"
 	"unsafe"
 
-	"github.com/appscode/go/log"
```
@ssmccoy let me know if you'd like this removed. I noticed that this import was already there since  originally being introduced in appscode fork.

jacksonrnewhouse commented on client/client.go:
```diff
@@ -115,13 +182,30 @@ func (client *Client) writeLoop() {
+		}
 
-		client.conn.Write(req.data[0])
+		if _, err := client.getConn().Write(req.data[0]); err != nil {
```
Can you factor this out into a function? It's repeated many times. Something like
```
(client * Client) WriteBytes(bytes []byte) err{
		if _, err := client.getConn().Write(bytes); err != nil {
			if err = client.reconnect(err); err != nil {
                                   return err
			}
		}
}

toli-belo replied:
refactored to use a one-liner

jacksonrnewhouse commented on client/client.go:
```diff
@@ -136,49 +220,74 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 
-	// TODO I doubt this error handling is right because it looks
-	// really complicated.
+func (client *Client) reconnect(err error) error {
```
I think some sort of atomic reconnection control makes sense, but this combined with the grabReconnectState() makes it pretty hard to reason about. Can you lay out the reconnect semantics we expect? What do we want to have happen if two reconnect() calls are invoked concurrently?

toli-belo replied:
@jacksonrnewhouse 
1. Lock is to prevent the other of the 2 threads from using the connection for actual send/receive.
2. grabReconectState() helps resolve the race between the two threads if both failed at (nearly) same time and attempting to reconnect.

toli-belo replied:
update: removed client Lock, changed read/write loop reconnect behavior to both exit and look at conn==nil as indicator that reconnection is in progress.

jacksonrnewhouse commented on client/client.go:
```diff
@@ -289,20 +401,26 @@ func (client *Client) request() *request {
+func (client *Client) submit(pt rt.PT, funcname string, payload []byte) (handle string, err error) {
 
-	client.outbound <- client.request().submitJob(pt, funcname, IdGen.Id(), payload)
+	defer func() {
```
This function is now present at least 3 times. Can we refactor it?

toli-belo replied:
Good idea, I looked around but it looks like it's the shortest form so far I can think of where (a) we are deferring a function (b) which calls recover(), (c) assigns named return variables (e.g. handle and err in this case), and (d) one of the return variables default values differs between "" (emptry string), or nil, but let me know if you've seen a shorter construct though.

jacksonrnewhouse commented on client/response.go:
```diff
@@ -56,9 +56,9 @@ func (resp *Response) Update() (data []byte, err error) {
-	data := bytes.SplitN(resp.Data, []byte{'\x00'}, 2)
-	if len(data) != 2 {
-		err = fmt.Errorf("Invalid data: %v", resp.Data)
+	data := bytes.Split(resp.Data, []byte{'\x00'})
```
This seems like a regression. the data after the first `'\x00'` could have null bytes, and this change would remove them from data[1] which gets called below.

toli-belo replied:
@jacksonrnewhouse , or interesting, the previous version SplitN was failing with the example client's requests because they contained I believe 3 elements after split, which one is right?

toli-belo replied:
reverted to original version, appears to have been a fluke on my side.

jacksonrnewhouse commented on worker/worker.go:
```diff
@@ -186,17 +186,15 @@ func (worker *Worker) Ready() (err error) {
-			worker.addFunc(funcname, f.timeout)
-		}
-	})
+	//	worker.once.Do(func() {
```
Rather than commenting out, delete and rely on git to recover if we decide we want it again.

toli-belo replied:
fixed, this wasn't meant to be left in the file.

jacksonrnewhouse reviewed (changes requested):
I gave it a pass and raised some issues.

Commit:
Address PR comments.

* Revert Split to SplitN etc in response.go
* Introduce a writeReconnectCleanup function with a compact write/reconnect/cleanup op
* Add a todo for a more sophisticated gearman connection health checker
* Remove  commented out code

toli-belo reviewed (commented):
made a few changes to address comments

Commit:
Change usages of log in worker and client to new logHandler, make it optional to pass

Also rename some handles to match a common pattern such as [sS]omethingHandler

Commit:
Fix writeLoop exit on reconnect

Commit:
Work around using the Lock in client, optimize read and write loop reconnect behavior

toli-belo reviewed (commented):
made a few updates based on yours guys comments.

Commit:
Streamline locking and new job grab/requesting

Add current jobs count for diagnostics

ssmccoy commented on client/client.go:
```diff
@@ -32,53 +31,128 @@ type connection struct {
+	connVersion int
 }
 
+type chanStruct struct {
```
I'm not sure how `Struct` adds anything to the name of this thing. Maybe just call it `chans` ? (Also why isn't it associated with `connection`?)

toli-belo replied:
removing chanStruct, moving channels back to connection.

ssmccoy commented on client/client.go:
```diff
@@ -136,77 +238,107 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
+func (client *Client) grabReconnectState() (success bool) {
+	client.reconnectLock.Lock()
+	defer client.reconnectLock.Unlock()
+	if client.reconnectingActive { // another thread is already reconnecting to server, this thread will exit
```
Regarding this comment, it's not necessarily a _thread_ that is already reconnecting — it could be a goroutine which is running on the same thread.

toli-belo replied:
that code is gone now

ssmccoy commented on client/client.go:
```diff
@@ -95,14 +189,21 @@ func (client *Client) writeLoop() {
+			client.requestPool.Put(req)
+		}
+
+		if exit := client.writeReconnectCleanup(cleanupHandle, []byte(rt.ReqStr)); exit {
```
Seems that you could just _not_ store `exit` here and just say...

```
if client.writeReconnectCleanup(...) {
    return
}
```

toli-belo replied:
got too excited about `go`, fixed this

ssmccoy commented on client/client.go:
```diff
@@ -87,6 +161,26 @@ func NewConnected(conn net.Conn) (client *Client) {
+	return client.conn != nil
+}
+
+func (client *Client) writeReconnectCleanup(cleanupHandle func(), ibufs ...[]byte) (exit bool) {
```
You declare `exit` as the rval but don't use it. Why not just declare the return type as `) bool {` ?

toli-belo replied:
fixed

ssmccoy commented on client/client.go:
```diff
@@ -32,53 +31,128 @@ type connection struct {
+			logHandler(Info, fmt.Sprintf("Trying to connect to server %v ...", addr))
+		}
+		for {
+			for num_tries := 1; ; num_tries++ {
```
It's more canonical on go to use `camelCase` rather than `snake_case`

toli-belo replied:
fixed

ssmccoy commented on client/client.go:
```diff
@@ -32,53 +31,128 @@ type connection struct {
+type LogHandler func(level LogLevel, message ...string)
+
+func (client *Client) Log(level LogLevel, message ...string) {
+	if client.logHandler != nil {
```
Rather than repeating this `nil` check throughout the code why not do it once?

```
if logHandler == nil {
    logHandler = func(level LogLevel, message ...string) {}
}

toli-belo replied:
good idea, fixed

ssmccoy commented on client/client.go:
```diff
@@ -32,53 +31,128 @@ type connection struct {
+			conn, err = net.Dial(network, addr)
+			if err != nil {
+				// looks like there is another problem, go back to the main loop
+				time.Sleep(time.Second)
```
Does this mean you'll sleep and then sleep again?

Also, I'm curious, does `net.Dial()` ever return `nil` for `conn`? Won't that cause a segfault?

toli-belo replied:
we connect (first time), we know that gearman is up, then we disconnect, sleep again, and dial, but if connection fails again then we sleep and go back to square 1.

ssmccoy commented on client/client.go:
```diff
@@ -32,53 +31,128 @@ type connection struct {
 
-// Return a new client from an established connection. Largely used for
-// testing, though other use-cases can be imagined.
-func NewConnected(conn net.Conn) (client *Client) {
```
I would try not to change the signature of exported methods unless you have a really good reason. It breaks API compatibility.

Presently this is a drop-in replacement for `github.com/mikespook/gearman-go` and that will no longer be the case now.

toli-belo replied:
fixed: implemented a replacement for the original `NewConnected`

ssmccoy commented on client/client.go:
```diff
@@ -95,14 +189,21 @@ func (client *Client) writeLoop() {
+	for req := range client.channels.outbound {
 
-		// todo handle errors.
+		cleanupHandle := func() {
```
This closure doesn't seem necessary. You could just pass the `req` into the `writeReconnectCleanup` method and rather than calling the closure just call `client.requestPool.put(req)` at that point.

toli-belo replied:
done that

ssmccoy commented on client/client.go:
```diff
@@ -136,77 +238,107 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
-	if client.conn != nil {
-		return nil
+func (client *Client) grabReconnectState() (success bool) {
+	client.reconnectLock.Lock()
```
I am not clear why you would use a mutex here instead of CAS. You are doing a compare and swap in this operation, why not push it down to the memory controller by using `sync/atomic.CompareAndSwapUnit32`? i.e.

```
func (client *Client) lockReconnect() bool {
    return atomic.CompareAndSwapUint32(&client.reconnecting, 0, 1)
}
```

This has the same behavior and is less expensive (and less code).

toli-belo replied:
fixed.

ssmccoy commented on client/client.go:
```diff
@@ -136,77 +238,107 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
+		return errors.New("Was expecting nil when replacing with new connection")
 	}
 
+	_ = atomic.SwapPointer((*unsafe.Pointer)(unsafe.Pointer(&client.channels)),
```
I am not sure this `SwapPointer` is doing much here.

Also, since it always creates a new `chanStruct` when the `connection` is replaced, why not just place the channels in the `connection` object?

toli-belo replied:
moving them into connection will require checking if client.conn==nil, but not necessary since they are being closed anyway, and they have to be closed in the current design, so moving them into `client` seems cheaper.

ssmccoy commented on client/client.go:
```diff
@@ -136,77 +238,107 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
+	var err error
+	if n, err = io.ReadFull(conn, buf); err != nil {
+		go client.reconnect(err)
+		return 0, true
```
Is `0` necessarily true and should it be?

toli-belo replied:
ditto

ssmccoy commented on client/client.go:
```diff
@@ -136,77 +238,107 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 }
 
+func (client *Client) readReconnect(buf []byte) (n int, exit bool) {
+	conn := client.conn
```
Typically when using `atomic.CompareAndSwapPointer` it's a good idea to also using `atomic.LoadPointer` as well as it properly implements memory fencing.

toli-belo replied:
done

ssmccoy commented on client/client.go:
```diff
@@ -282,27 +416,35 @@ func (client *Client) process(resp *Response) {
+		if e := recover(); e != nil {
+			handle = ""
+			err = e.(error)
+		}
```
It's not really considered best practice to rely on `recover`, also this is an unsafe typecast here it _could_ error as there is no guarantee that `panic()` is called with an `error` object.

toli-belo replied:
added cast checking and moved to another function

ssmccoy reviewed (changes requested):
Here are my comments on the client portion, I will go over the worker portion next — this is a very large change, it would be a bit easier to review if it were broken up into smaller changes.

ssmccoy commented on worker/agent.go:
```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
 	defer func() {
 		if err := recover(); err != nil {
-			a.worker.err(err.(error))
+			a.reconnect_error(err.(error))
```
I mentioned this earlier in the review but you should use `camelCase` because that's the go way.

toli-belo replied:
fixed

ssmccoy commented on worker/agent.go:
```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
 }
 
-func (a *agent) grab() {
+func (a *agent) grab() (err error) {
```
Please don't define variables without using them (i.e. this should be `grab() error {` since you never use `err`)

toli-belo replied:
fixed

ssmccoy commented on worker/agent.go:
```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
+		if len(leftdata) > 0 { // some data left for processing
+			data = append(leftdata, data...)
+		}
+		if len(data) < rt.MinPacketLength { // not enough data
```
You could remove this whole routine if you want to make a significant improvement to this system. :-) It's really silly and almost the same as the parsing in the client.

toli-belo replied:
let's put it in a new ticket

ssmccoy commented on worker/agent.go:
```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
+	startConnVersion := a.connectionVersion
+
+	// exit the loop if connection has been replaced because reconnect will launch a new work() thread
+	for startConnVersion == a.connectionVersion && !a.worker.isShuttingDown() {
```
Why not just look at the [pointer `a.conn`](https://play.golang.org/p/lw-kjSIbP4-) points to instead of keeping this `connectionVersion` value around?

toli-belo replied:
good point, switched to use `agent.rw` which is currently atomically switched.

ssmccoy commented on worker/agent.go:
```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
-	conn, err := net.Dial(a.net, a.addr)
-	if err != nil {
-		return err
+func (a *agent) grabReconnectState() (success bool) {
```
I also think this should use `atomic.CompareAndSwapUint32()` instead of this double-lock/compare stuff

toli-belo replied:
implemented `lockReconnect()` like in client

ssmccoy commented on worker/agent.go:
```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
+
+		a.worker.Log(Info, "Successfully Connected to:", a.addr)
+
+		newRw := bufio.NewReadWriter(bufio.NewReader(a.conn), bufio.NewWriter(a.conn))
```
These buffers are unnecessary and redundant? The I/O on a socket is already buffered. :-/

toli-belo replied:
new ticket

ssmccoy commented on worker/agent.go:
```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
+			continue
+		}
+
+		_ = conn.(*net.TCPConn).SetKeepAlive(true)
```
I would really prefer this didn't cast without checking, and and used `net.DialTCP()` instead of `net.Dial()` then casting to a `TCPConn` without first checking the safety of it.

toli-belo replied:
seems not having effect on ELBs anyway, removing this.

Commit:
Make improvements based on PR feedback.

* move channels back into client
* restore NewConnected constructor
* add safe error casting for recover output
* atomic loads for conn

Commit:
Update on PR comments, load atomics

toli-belo reviewed (commented):
more feedback changes, restored some original constructor function names

Commit:
Move channel close to go after conn close

Commit:
Add conn comparison in read loop

Commit:
Make sure write loop checks connection before series of writes for a single transmission

This is to help the integrity of the writes in case reconnects occur mid-transmission

Commit:
Change some names to camelCase

ssmccoy commented on client/client.go:
```diff
@@ -136,77 +251,105 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
+		client.Log(Warning, fmt.Sprintf("Non-fatal error %v, while closing connection to %v", closeErr, client.addr))
+	}
 
+	close(client.expected)
```
I would still close these *after* changing them.

Otherwise I think we're ready to go.

Commit:
Close outbound and expected after they are replaced, make access atomic.

Commit:
Move channel close back up before reconnecting

Reason is we want any hung up submit, echo, status callers to
error out before we attempt reconnecting because it might be minutes
before the gearman is back up and we don't want them hanging.

ssmccoy commented on client/client.go:
```diff
@@ -200,7 +208,7 @@ func (client *Client) writeLoop() {
 	// goroutines to all write as quickly as possible, uses a channel and the
 	// writeLoop lives in a separate goroutine.
-	for req := range client.outbound {
+	for req := range client.loadChans().outbound {
```
Worth noting that the expression in [range is not recalculated](https://play.golang.org/p/_P33M1fLCIX)

toli-belo replied:
yeah we don't want it to recalculate, we want it to bounce out of this loop when it's done
